### PR TITLE
Separate diagnostics from rule results

### PR DESCRIPTION
**Pull request by cocreature**
This is a simplified version of #1060. The main change is that it leaves filepaths mostly untouched since we can do that later once we have a decent test suite that covers all of the edge cases but there are also quite a few other changes and simplifications.

I tested it on Linux and Windows both manually and via test suites.

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/docs/source/support/release-notes.rst), if appropriate

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs [hunk 1]
```diff
@@ -167,20 +161,61 @@ prettyDiagnostic (fp, LSP.Diagnostic{..}) =
     where
         sev = fromMaybe LSP.DsError _severity
 
-prettyDiagnosticStore :: DiagnosticStore -> Doc SyntaxClass
```
Comment (cocreature) [inline compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: These functions were simply unused.

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs [hunk 2]
```diff
@@ -96,28 +112,6 @@ defDiagnostic _range _message = LSP.Diagnostic {
   , _relatedInformation = Nothing
   }
 
-filterSeriousErrors ::
-    FilePath ->
-    [LSP.Diagnostic] ->
-    [LSP.Diagnostic]
-filterSeriousErrors fp =
```
Comment (cocreature) [inline compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: Also unused.

File: daml-foundations/daml-tools/language-server-tests/src/tests/issue-reports.ts [hunk 3]
```diff
@@ -1,182 +0,0 @@
-// Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
```
Comment (cocreature) [inline daml-foundations/daml-tools/language-server-tests/src/tests/issue-reports.ts]: This test broke due to the fact that we no longer emit empty sets of diagnostics. Given that this does not test anything LSP specific and we test the same codepath in daml-ghc-shake-test-ci deleting it seems like a better option than trying to fix the test suite given that we should hopefully be able to remove all of it soon.

File: daml-foundations/daml-tools/language-server-tests/tests/diagnostics/library-modules/EXPECTED.json [hunk 4]
```diff
@@ -1,7 +0,0 @@
-{
```
Comment (cocreature) [inline daml-foundations/daml-tools/language-server-tests/tests/diagnostics/library-modules/EXPECTED.json]: Removed for the same reason.

File: daml-foundations/daml-tools/language-server-tests/tests/diagnostics/multi-module/Libs/A.daml [hunk 5]
```diff
@@ -1,10 +0,0 @@
--- Copyright (c) 2019, Digital Asset (Switzerland) GmbH and/or its affiliates.
```
Comment (cocreature) [inline daml-foundations/daml-tools/language-server-tests/tests/diagnostics/multi-module/Libs/A.daml]: Removed for the same reason.

File: daml-foundations/daml-ghc/test-src/DA/Test/GHC.hs [hunk 6]
```diff
@@ -207,7 +207,7 @@ checkDiagnostics log expected got = do
             (\expFields -> not $ any (\diag -> all (checkField diag) expFields) got)
             expected
     pure $ if
-      | length expected /= length got -> Just $ "Wrong number of diagnostics, expected " ++ show (length expected)
+      | length expected /= length got -> Just $ "Wrong number of diagnostics, expected " ++ show (length expected) ++ ", but got " ++ show (length got)
```
Comment (associahedron) [inline daml-foundations/daml-ghc/test-src/DA/Test/GHC.hs]: This line could do with breaking up into multiple lines.

File: compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs (hunk 1 above)
Comment (neil-da) [reply compiler/haskell-ide-core/src/Development/IDE/Types/Diagnostics.hs]: Yes, I had noticed this, but wondered if DMD was going to use them in a future version...

File: compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs [hunk 7]
```diff
@@ -213,18 +205,21 @@ getValues :: forall k v. IdeRule k v => Var Values -> k -> FilePath -> IO (Maybe
...
 -- | Open a 'IdeState', should be shut using 'shakeShut'.
 shakeOpen :: (Event -> IO ()) -- ^ diagnostic handler
           -> Logger.Handle
           -> ShakeOptions
           -> Rules ()
           -> IO IdeState
-shakeOpen diags shakeLogger opts rules = do
-    shakeExtras <- ShakeExtras diags shakeLogger <$> newVar Map.empty <*> newVar Map.empty
+shakeOpen eventer logger opts rules = do
+    shakeExtras <- do
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs]: I really like this style of code.

File: compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs [hunk 8]
```diff
@@ -263,20 +258,22 @@ useStale IdeState{shakeExtras=ShakeExtras{state}} k fp =
...
+unsafeClearAllDiagnostics IdeState{shakeExtras = ShakeExtras{diagnostics}} =
+    writeVar diagnostics emptyDiagnostics
 
 -- | Clear the results for all files that do not match the given predicate.
 garbageCollect :: (FilePath -> Bool) -> Action ()
 garbageCollect keep = do
-    ShakeExtras{state} <- getShakeExtras
-    liftIO $ modifyVar_ state $ return . Map.filterWithKey (\file _ -> keep file)
+    ShakeExtras{state, diagnostics} <- getShakeExtras
+    liftIO $
```
Comment (neil-da) [inline compiler/haskell-ide-core/src/Development/IDE/State/Shake.hs]: I'd have put the do on the same line as the `liftIO`. I find the below indentation a bit weird.

**Pull request merged**